Fix default egress rule deletion logic in case of an egress rule definition

mustafasencer opened the pull request:
**Issue:** https://github.com/aws-controllers-k8s/community/issues/1604

**Description of changes:**
If the user defines an egress rule which is the same as the “default“ egress rule (defined implicitly by AWS), then the ec2-controller raises the below error. Although the SG is created in AWS, the controller is unable to sync the resource.
```txt
2023-01-26T16:19:39.988Z    DEBUG    ackrt    <<<<< rm.syncSGRules    {"account": "647927084307", "role": "", "region": "eu-west-1", "kind": "SecurityGroup", "namespace": "ack-system", "name": "ack-sg-test", "is_adopted": false, "generation": 1}
2023-01-26T16:19:39.988Z    DEBUG    ackrt    <<<< rm.sdkCreate    {"account": "647927084307", "role": "", "region": "eu-west-1", "kind": "SecurityGroup", "namespace": "ack-system", "name": "ack-sg-test", "is_adopted": false, "generation": 1, "error": "InvalidPermission.Duplicate: the specified rule \"peer: 0.0.0.0/0, ALL, ALLOW\" already exists\n\tstatus code: 400, request id: 160a5101-48d5-4de3-afc9-7ff6d7b99f16"}
2023-01-26T16:19:39.988Z    DEBUG    ackrt    <<< rm.Create    {"account": "647927084307", "role": "", "region": "eu-west-1", "kind": "SecurityGroup", "namespace": "ack-system", "name": "ack-sg-test", "is_adopted": false, "generation": 1, "error": "InvalidPermission.Duplicate: the specified rule \"peer: 0.0.0.0/0, ALL, ALLOW\" already exists\n\tstatus code: 400, request id: 160a5101-48d5-4de3-afc9-7ff6d7b99f16"}
2023-01-26T16:19:39.988Z    DEBUG    ackrt    << r.createResource    {"account": "647927084307", "role": "", "region": "eu-west-1", "kind": "SecurityGroup", "namespace": "ack-system", "name": "ack-sg-test", "is_adopted": false, "generation": 1, "error": "InvalidPermission.Duplicate: the specified rule \"peer: 0.0.0.0/0, ALL, ALLOW\" already exists\n\tstatus code: 400, request id: 160a5101-48d5-4de3-afc9-7ff6d7b99f16"}
```
The change in this PR just changes the order of the flow executing the deletion of the “default“ egress rule in case an egress rule is defined by the user. Because the `syncSGRules` during creation does not really sync the SG rules as it does  not have a `latest` state during performing the sync, the “default“ egress rule hangs there and basically causes the issue.

**Note:** I am not sure whether this case requires a specific e2e test case; the current `test_rules_create_update_delete` case covers the situation where only an ingress rule is defined and then it only patches the egress rule definition which does not 100% cover this situation, because this particular case only pops up during the creation step where the default egress rule check occurs. In order to cover it, the egress rule should be defined from the beginning on I suppose. Just wanted to share my limited perspective 😄 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

A-Hilaly commented:
Thank you @mustafasencer ! QQ: What if the default egress rule was already deleted, should ignore 404 NotFound errors?

A-Hilaly commented:
/ok-to-test

jljaco commented:
Thanks for doing this!  I have created a new e2e test that hopefully will validate this case in 

https://github.com/aws-controllers-k8s/ec2-controller/pull/124

how about we get that one merged and then confirm that the test passes for this change?

jljaco commented:
/hold

jljaco commented:
> Thanks for doing this! I have created a new e2e test that hopefully will validate this case in
> 
> #124
> 
> how about we get that one merged and then confirm that the test passes for this change?

@mustafasencer #124 is now merged.  In your branch, would you please remove the `@pytest.mark.xfail` line [here](https://github.com/aws-controllers-k8s/ec2-controller/blob/c027a797a7b93a1b3272f8d73097738ca8912153/test/e2e/tests/test_security_group.py#L162)?  That way we can use the new test to confirm that your fix works.

jljaco commented:
/retest

LikithaVemulapalli commented:
/retest

mustafasencer commented:
@jljaco Thanks for the prompt implementation of the e2e test case. I noticed that the e2e test cases were checking the case where the `spec.egressRules` needed to be updated (due to the “default“ egress rule) after the resource has been synced. Hence, I introduced a late initialisation logic there to account for the `spec.egressRules` field's population after the readOne operation. The e2e test cases are fully covered with this change. Feel free to leave your comments or update the PR directly, as I am not 100% familiar with the codebase, thus have perhaps missed a crucial part.

@A-Hilaly Although the case you mentioned should not be expected, nevertheless I added the ignore logic for the `InvalidPermission.NotFound` in case the default egress rule is not present.

Thanks!

mustafasencer commented:
/retest

A-Hilaly commented on pkg/resource/security_group/manager.go:
```diff
@@ -257,7 +257,12 @@ func (rm *resourceManager) lateInitializeFromReadOneOutput(
+	if observedKo.Spec.EgressRules != nil && latestKo.Spec.EgressRules == nil {
+		latestKo.Spec.EgressRules = observedKo.Spec.EgressRules
+	}
+	return &resource{latestKo}
```
Can you please include the files you modified to generate this code? `generator.yaml` and what changed in `apis/v1alpha1`

mustafasencer replied:
Ohh, sorry I completely forgot about the generation part during manual testing. Pushing the generated code quick...

A-Hilaly reviewed (commented):
Looks great, thank you @mustafasencer! I have one comment bellow

A-Hilaly reviewed (approved):
Neat! Thank you @mustafasencer !

jljaco commented:
/lgtm

Thank you @mustafasencer !

jljaco commented:
/lgtm

jljaco commented:
Thank you @mustafasencer

jljaco commented:
/unhold
